executor/aggregate.go

```diff
@@ -45,39 +96,484 @@ type HashAggExec struct {
+			return
+		}
+		needShuffle = true
+		if chk.NumRows() == 0 {
+			return
+		}
+		w.giveBackCh <- &HashAggInput{
+			chk:        chk,
+			giveBackCh: w.inputCh,
+		}
+	}
+}
+
+func (w *HashAggPartialWorker) updateIntermData(ctx sessionctx.Context, sc *stmtctx.StatementContext, chk *chunk.Chunk, finalConcurrency int) (isOk bool) {
+	inputIter := chunk.NewIterator4Chunk(chk)
+	for row := inputIter.Begin(); row != inputIter.End(); row = inputIter.Next() {
+		groupKey, err := w.getGroupKey(sc, row)
+		if err != nil {
+			w.globalOutputCh <- &AfFinalResult{err: errors.Trace(err)}
+			return false
+		}
+		finalWorkerIdx := int(murmur3.Sum32(groupKey)) % finalConcurrency
+		if w.aggCtxsMaps[finalWorkerIdx] == nil {
+			w.aggCtxsMaps[finalWorkerIdx] = make(aggCtxsMapper, 0)
+			w.groupSets[finalWorkerIdx] = mvmap.NewMVMap()
+		}
+		aggCtxMap, groupSet := w.aggCtxsMaps[finalWorkerIdx], w.groupSets[finalWorkerIdx]
+		if len(groupSet.Get(groupKey, w.groupVals[:0])) == 0 {
+			groupSet.Put(groupKey, []byte{})
+		}
+		aggEvalCtxs := w.getContext(sc, groupKey, aggCtxMap)
+		for i, af := range w.aggFuncs {
+			err = af.Update(aggEvalCtxs[i], sc, row)
+			if err != nil {
+				w.globalOutputCh <- &AfFinalResult{err: errors.Trace(err)}
+				return false
+			}
+		}
+	}
+	return true
+}
+
+// shuffleIntermData shuffles the intermediate data of partial workers to corresponded final workers.
+// We only support parallel execution for single-machine, so process of encode and decode can be skipped.
+func (w *HashAggPartialWorker) shuffleIntermData(sc *stmtctx.StatementContext) {
+	for i := range w.groupSets {
+		if w.aggCtxsMaps[i] == nil {
+			continue
+		}
+		w.outputChs[i] <- &HashAggIntermData{
+			groupSet:    w.groupSets[i],
+			groupCtxMap: w.aggCtxsMaps[i],
+		}
+		w.aggCtxsMaps[i] = nil
+	}
+}
+
+// getGroupKey evaluates the group items and args of aggregate functions.
+func (w *HashAggPartialWorker) getGroupKey(sc *stmtctx.StatementContext, row chunk.Row) ([]byte, error) {
+	w.groupValDatums = w.groupValDatums[:0]
+	for _, item := range w.groupByItems {
+		v, err := item.Eval(row)
+		if item.GetType().Tp == mysql.TypeNewDecimal {
+			v.SetLength(0)
+		}
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		w.groupValDatums = append(w.groupValDatums, v)
+	}
+	var err error
+	w.groupKey, err = codec.EncodeValue(sc, w.groupKey[:0], w.groupValDatums...)
+	return w.groupKey, errors.Trace(err)
+}
+
+func (w baseHashAggWorker) getContext(sc *stmtctx.StatementContext, groupKey []byte, mapper aggCtxsMapper) []*aggregation.AggEvaluateContext {
+	aggCtxs, ok := mapper[string(groupKey)]
+	if !ok {
+		aggCtxs = make([]*aggregation.AggEvaluateContext, 0, len(w.aggFuncs))
+		for _, af := range w.aggFuncs {
+			aggCtxs = append(aggCtxs, af.CreateContext(sc))
+		}
+		mapper[string(groupKey)] = aggCtxs
+	}
+	return aggCtxs
+}
+
+func (w *HashAggFinalWorker) run(ctx sessionctx.Context, waitGroup *sync.WaitGroup) {
```

Comment:
Please split this function to two functions:
- one is used to receive partial results from partial workers and update the final result.
- another one is used to calculate the final results and return them to the main thread.